Add claims list endpoint

austin-martinez opened the pull request:
Open questions this PR will not address
- Pagination/sorting/filtering
- How we handle caching/persisting claims

Questions this PR could address
- Where does the EVSS response get translated to a model (do we even need to translate to a model? this seems dependent on caching/persisting answer)
- What attributes from EVSS do we need to pass through to the frontend

austin-martinez committed:
Add claims list endpoint

greggersh commented on app/models/claim.rb:
```diff
@@ -1,3 +1,20 @@
 class Claim < ActiveModelSerializers::Model
   attr_accessor :id
+
+  def self.fetch_all(headers)
```
I think this is all OK for now, but what I think we should evolve towards is a Veteran model that is stored in the session with the information on the Veteran (maybe that should be a User model, since not all Users will be Veterans), and a method on the Veteran model to get all claims, which dispatches the call to the EVSS class to do the work. I'm trying to think how we'd call this if we were building this app as a plain ole' rails app, in which case we'd have a User instance instantiated with the session, and we'd say something like `current_user.claims` to get their claim list. I think we want to mimic that as closely as possible. Thoughts?

carlmjohnson replied:
Thinking about it in terms of separation of concerns, `EVSS:ClaimsService` shouldn't care about the user model. It just needs VAAFI headers from some place in order to make its requests to the backing service. The Claim model probably also shouldn't care about the details of how claims are authenticated/authorized, since it's just a translation layer that turns the backend JSON into something suitable for the front end, possibly by making multiple calls or flattening or nesting bits of JSON from the backend. So, that leaves a User model and the Controller. When a user gets authenticated by ID.me and we figure out their VAAFI headers by going through MVI, we should store that as User.vaafi_headers, and then the controller's job is to look at request, see that it has a session.user_id and then connect that User's vaafi_headers to the various EVSS services by way of the Claim model.

austin-martinez replied:
I updated this to show how it would look with the User model. I like the simplicity of being able to just call `current_user.claims` in the controller, but calling `Claim.fetch_all(current_user.vaafi_headers)` directly is also an option if we want to avoid bloating the User model

greggersh commented:
Let's pass everything for now, for transparency, we can peel that back as we go.

I'm OK with no model other than JSON just being passed around for now.

Other than the broader comment above, this looks good!

carlmjohnson commented on spec/client_spec.rb:
```diff
@@ -18,7 +18,7 @@
       'va_eauth_middleName'=> 'A',
       'va_eauth_authenticationmethod'=> 'DSLogon',
-      'va_eauth_assurancelevel'=> '2',
+      'va_eauth_assurancelevel'=> '2'
```
Is this preferred in Ruby? In JavaScript, I know you have to omit the trailing comma because of older browsers, but in Python I always had a standard of including trailing commas so that diffs are cleaner when you add or remove a line. If the Ruby standard is to match up with JS, I'll do that.

austin-martinez replied:
This is something Rubocop complained about. I've always omitted trailing commas but could go either way on it, so for now we should just favor the Rubocop rules

carlmjohnson replied:
Yeah, 🚫 🚲 shedding.

carlmjohnson commented on spec/requests/claims_spec.rb:
```diff
@@ -1,8 +1,8 @@
-  it "creates a Claim" do
-    post "/claims", params: { claim: { id: 1 } }
-    expect(response).to match_response_schema("claim")
+  it "lists all Claim" do
```
We can probably make this DRY by looping through all the methods. Also, we should add methods for all the endpoints in the WADL file.

carlmjohnson replied:
Oops, I didn't mean here, but in client_spec.rb.

earthboundkid commented:
> I'm OK with no model other than JSON just being passed around for now.

Yes, we can try to practice YAGNI for now, but I do want to add a small note of caution since I just came off a project at PBS where we had a backend layer get passed to the front end with very little translation and it was a nightmare, since two teams had to do coordinate releases for even minor changes. I think what Austin has here is a good compromise of simplicity with room for future improvements.

austin-martinez committed:
Add user model to handle passing headers

austin-martinez commented:
I think my statement about just passing back the JSON directly from EVSS was a bit premature, if we want to follow JSON API specs we definitely need to manipulate it somewhere, translating to a model is the simplest way as we can then just let ActiveModel::Serializers do its magic instead of manually creating a hash that conforms to the JSON API spec

earthboundkid commented:
This looks really good. Other than the merge conflicts, 👍 .

austin-martinez committed:
Merge remote-tracking branch 'origin/master' into claims-index-api

Conflicts:
	lib/evss.rb
	spec/client_spec.rb

austin-martinez committed:
Use cassette in spec

austin-martinez committed:
Remove expectation of response body being a string

Using the Faraday ParseJSON middleware parses the response string into a Hash/Array

earthboundkid commented:
I got an error when I curled /claims. It looks like you need to add `require "evss"` to claim.rb. Also, can you clean up the rubocop double quote issues?

austin-martinez committed:
Require EVSS and Rubocop fixes
